### ISSUE TITLE
Possible to use this to bypass SNI based https censorship?

### ISSUE THREAD

**Issue by aaomidi** [elazarl/goproxy#308]
Hi!

SNI is the field that most countries such as Iran, Egypt, China, etc use to block specific websites without having to resort to IP blocking.

For example, in Iran, google.com works while youtube.com is blocked.

[Since Google doesn't actually use the SNI field](https://hal.inria.fr/hal-01202712/document), I was thinking I can delete/spoof that value using this proxy.

Is that possible? Can you please point me towards the right direction?

**Comment by elazarl** [comment]
Hmmm...
I think you need to intercept HTTPS traffic (MITM), and then make sure to
change the SNI using Go's http.Transport.

I'm not sure how to do that in Go's side.

**Comment by elazarl** [comment]
I see, in that case you have to intercept the actual bit stream of the tcp
connection. Maybe a TCP proxy would be a better fit? Still possible

**Comment by aaomidi** [comment]
I think what I can probably do is get GoLang to reconstruct the TCP packets (DPI the data stream) and then reconstruct HTTP packets out of them.

Going to be a messy solution though :/

**Comment by eternal-flame-AD** [comment]
Directly modifying the handshake won't work as in the final stage of the handshake the hash of the whole handshake would be used to generate the encryption keys. See [rfc5246](https://tools.ietf.org/html/rfc5246#section-7.4.9) for details.

I have written a MITM to circumvent the SNI censorship on Pixiv for China and I think it works for other sites as well. https://github.com/eternal-flame-AD/go-pixiv

**Comment by ErikPelli** [comment]
I agree with the other answers, this isn't the right project to implement stuff like this.
Closing this issue.
